x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -140,165 +144,122 @@ public String sessionId() {
+        PlanRunner runner,
         ActionListener<Result> listener
     ) {
-        PhysicalPlan physicalPlan = logicalPlanToPhysicalPlan(optimizedPlan(firstPhase), request);
-        runPhase.accept(physicalPlan, listener.delegateFailureAndWrap((next, result) -> {
+        PlanTuple tuple = subPlanIterator.next();
+
+        runner.run(tuple.physical, listener.delegateFailureAndWrap((next, result) -> {
             try {
                 profileAccumulator.addAll(result.profiles());
-                LogicalPlan newMainPlan = optimizedPlan(Phased.applyResultsFromFirstPhase(mainPlan, physicalPlan.output(), result.pages()));
-                LogicalPlan newFirstPhase = Phased.extractFirstPhase(newMainPlan);
-                if (newFirstPhase == null) {
-                    PhysicalPlan finalPhysicalPlan = logicalPlanToPhysicalPlan(newMainPlan, request);
-                    runPhase.accept(finalPhysicalPlan, next.delegateFailureAndWrap((finalListener, finalResult) -> {
+                LocalRelation resultWrapper = resultToPlan(tuple.logical, result);
+
+                // replace the original logical plan with the backing result
+                final PhysicalPlan newPlan = plan.transformUp(FragmentExec.class, f -> {
+                    LogicalPlan frag = f.fragment();
+                    return f.withFragment(
+                        frag.transformUp(
+                            InlineJoin.class,
+                            ij -> ij.right() == tuple.logical ? InlineJoin.inlineData(ij, resultWrapper) : ij
+                        )
+                    );
+                });
+                if (subPlanIterator.hasNext() == false) {
+                    runner.run(newPlan, next.delegateFailureAndWrap((finalListener, finalResult) -> {
                         profileAccumulator.addAll(finalResult.profiles());
                         finalListener.onResponse(new Result(finalResult.schema(), finalResult.pages(), profileAccumulator, executionInfo));
                     }));
                 } else {
-                    executePhased(profileAccumulator, newMainPlan, request, executionInfo, newFirstPhase, runPhase, next);
+                    // continue executing the subplans
+                    executeSubPlan(profileAccumulator, newPlan, subPlanIterator, executionInfo, runner, next);
                 }
             } finally {
                 Releasables.closeExpectNoException(Releasables.wrap(Iterators.map(result.pages().iterator(), p -> p::releaseBlocks)));
             }
         }));
     }
 
+    private LocalRelation resultToPlan(LogicalPlan plan, Result result) {
+        List<Page> pages = result.pages();
+        List<Attribute> schema = result.schema();
+        // if (pages.size() > 1) {
+        Block[] blocks = SessionUtils.fromPages(schema, pages);
+        return new LocalRelation(plan.source(), schema, LocalSupplier.of(blocks));
+    }
+
     private LogicalPlan parse(String query, QueryParams params) {
```

Comment:
@quux00 please double check this section.

Reply:
Yes, looks good. Same thing I did in my refactoring PR. As long as the various CrossCluster IT tests (like CrossClustersQueryIT) pass, you're good.